### PR TITLE
dialects: Update pdl dialect to leverage properties

### PR DESCRIPTION
**Pull request by martin-luecke**
This updates all operations of the PDL dialect to use properties rather than attributes. 
Correspondingly, the tests that relied on this were updated.
This also adds tests for `pdl.PatternOp`, which was untested before.

### CODE REVIEW

File: tests/dialects/test_pdl.py [hunk 1]
```diff
@@ -94,6 +97,49 @@ def test_build_operation_replace():
...
+
+        @Builder.implicit_region
+        def body() -> None:
+            pdl.RewriteOp(root.results[0])
+
+        pattern = pdl.PatternOp(1, "pattern", body)
+        pattern.verify()
+    assert "the pattern must contain at least one `pdl.operation`" in str(
+        verify_exception.value
+    )
```
Comment (superlopuh) [inline tests/dialects/test_pdl.py:140]:
```suggestion
    with pytest.raises(VerifyException, match="the pattern must contain at least one `pdl.operation`"):
        root = pdl.OperationOp("test.op")

        @Builder.implicit_region
        def body() -> None:
            pdl.RewriteOp(root.results[0])

        pattern = pdl.PatternOp(1, "pattern", body)
        pattern.verify()
```

and above

**Review by superlopuh — APPROVED** [review]
Thank you! Can you please make a couple of small changes to the tests?

File: tests/dialects/test_pdl.py (hunk 1 above)
Comment (martin-luecke) [reply tests/dialects/test_pdl.py]: I wondered if there was a better way, thanks!

**Review by math-fehr — APPROVED** [review]
Very nice!

**Pull request merged**